runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/smithy/Document.kt

```diff
@@ -4,161 +4,185 @@
+    }
 
     /**
-     * Returns the content as a float which may involve rounding
+     * Wraps a [kotlin.collections.List].
      */
-    val float: Float get() = content.toFloat()
+    data class List(val value: kotlin.collections.List<Document>) : Document() {
+        override fun toString() = value.joinToString(separator = ",", prefix = "[", postfix = "]")
+    }
 
     /**
-     * Returns the content as a double which may involve rounding
+     * Wraps a [kotlin.collections.Map].
      */
-    val double: Double get() = content.toDouble()
-
-    override fun toString(): String = content.toString()
-}
+    data class Map(val value: kotlin.collections.Map<kotlin.String, Document>) : Document() {
+        override fun toString() = value
+            .entries
+            .joinToString(
+                separator = ",",
+                prefix = "{",
+                postfix = "}",
+                transform = { (k, v) -> """"$k":$v""" }
+            )
+    }
 
-/**
- * Class representing document `array` type
- */
-data class SmithyArray(val content: List<Document>) : Document(), List<Document> by content {
     /**
-     * Returns [index] th element of an array as [SmithyNumber] if the element is of that type or null if not.
-     *
-     * @throws IndexOutOfBoundsException if there is no element with given index
+     * Represents a `null` value.
      */
-    fun getNumber(index: Int) = content[index] as? SmithyNumber
+    object Null : Document() {
+        override fun toString() = "null"
+    }
 
-    /**
-     * Returns [index] th element of an array as [SmithyBool] if the element is of that type or null if not.
-     *
-     * @throws IndexOutOfBoundsException if there is no element with given index
-     */
-    fun getBoolean(index: Int) = content[index] as? SmithyBool
+    fun asNumber() = (this as Number).value
+    fun asString() = (this as String).value
+    fun asBoolean() = (this as Boolean).value
+    fun asList() = (this as List).value
+    fun asMap() = (this as Map).value
 
-    /**
-     * Returns [index] th element of an array as [SmithyString] if the element is of that type or null if not.
-     *
-     * @throws IndexOutOfBoundsException if there is no element with given index
-     */
-    fun getString(index: Int) = content[index] as? SmithyString
+    fun asInt() = asNumber().toInt()
+    fun asByte() = asNumber().toByte()
+    fun asLong() = asNumber().toLong()
+    fun asFloat() = asNumber().toFloat()
+    fun asDouble() = asNumber().toDouble()
 
-    /**
-     * Returns [index] th element of an array as [SmithyArray] if the element is of that type or null if not.
-     *
-     * @throws IndexOutOfBoundsException if there is no element with given index
-     */
-    fun getArray(index: Int) = content[index] as? SmithyArray
+    val isNull: kotlin.Boolean
+        get() = this == Null
+
+    operator fun get(i: Int) = asList()[i]
+    operator fun get(i: kotlin.String) = asMap()[i]
```

Comment:
I don't know how I feel about all of these. We have casts here that can obviously fail. 

This needs further discussion I think.

Reply:
Isn't that just kind of a part-in-parcel risk of data being unstructured like this?

We could add the `getOrNull`-style methods as well.